### PR TITLE
Fix address balance overflow, reverse testnet address prefixes (like in kmd)

### PR DESCRIPTION
**Pull request by dimxy**
Fix address balance calculation overflow due to bigger amounts in kmd (removed MAX_MONEY limit for the balance var)
reverse transparent address prefixes for testnet (5<->0)

### CODE REVIEW

File: zebra-state/src/service/finalized_state/disk_format/transparent.rs [hunk 1]
```diff
@@ -242,7 +243,7 @@ impl AddressBalanceLocation {
         &mut self,
         unspent_output: &transparent::Output,
     ) -> Result<(), amount::Error> {
-        self.balance = (self.balance + unspent_output.value()).unwrap_or(MAX_MONEY.try_into()?);
+        self.balance = ( self.balance +  Amount::<MaxInt64>::try_from(i64::from(unspent_output.value())) )?;
```
Comment (DeckerSU) [inline zebra-state/src/service/finalized_state/disk_format/transparent.rs:246]:
```suggestion
self.balance = (self.balance + unspent_output.value().constrain::<MaxInt64>())?;
```
May be use of `constrain` looks better for conversion between different constraints of Amount? Non-principal, but just an offer.

Comment (dimxy) [reply zebra-state/src/service/finalized_state/disk_format/transparent.rs:246]: 'constrain' fits better here, 
thanks

**Comment by DeckerSU** [comment]
Fully synced `zebrad` with this PR changes and decided to check balances of Notary Nodes (as they are the most P2PK users in network) between Zebra `getaddressbalance` RPC and current KMD explorer. For this I wrote the following bash script:

[check-nn-balances-zebra.sh](https://github.com/DeckerSU/komodo_scripts/commit/b7ab15ac686cebbeabff80efcfac6b2d5d6564b6#diff-d41f0957a1fdf0ed099eebe90ba0aab6c274bb621a29e212a4fd86b00351f614)

As it turns out, for some nodes balances are not equal. Then, I guessed that balances are not equal only for nodes who have "in-fly" txes, mean, as NNs active doing splits, active notarizing, some of txes could be in mempool (i.e. just issued from node / just broadcasted and could be not affect balance on zebra or explorer yet). Then I made the same script for S3 notaries with guessing that S3 addresses are not active now and all of their txes are "settled down", and it give me almost equal picture for all nodes except few ones:
image
Like alien_AR, computergenie_NA ... But it turns out that operators of these nodes used the same addresses in S3 and S6, so, they are not equal at present bcz addresses are "very active".

So, parsing p2pk to legacy script changes looks good for me and I want to approve this PR (bcz it's logic and other things looks perfect), but I want somebody else to keep an eye open on this address balance differences. Probably we need some other tests. Like:

1. Create 1000 random addresses.
2. Send 500 txes on each with different values in sat. Like 0.00000051,  0.00000052, etc. (just more than dust threshold), half using P2PK and half using P2PKH script.
3. Compare these addresses balances after. If everything is Ok - than it means that zebra holding / checking / accounting balances correctly for P2PK and P2PKH both.

Do you like such idea?

**Comment by dimxy** [comment]
I think address balances basically should not be affected by txns in mempool but indeed, some active address balances may change during tests.
Actually in this PR we fixed overflow issues so I think it's good to create tests for big amounts too. I have a testnet config so I could make tests for address balances with many txns and both low and big amounts

**Comment by DeckerSU** [comment]
> I think address balances basically should not be affected by txns in mempool but indeed, some active address balances may change during tests. Actually in this PR we fixed overflow issues so I think it's good to create tests for big amounts too. I have a testnet config so I could make tests for address balances with many txns and both low and big amounts

It depends on implementation. I didn't remember how it's implemented in komodod with addressindexes, etc. I mean, let's imagine user spent 0.00001 KMD ... what addressbalance will be returned by daemon? Actual balance before spend, as tx is not confirmed, or after? It's important. Bcz different implementation in zebra and in komodod could give different results. Actually, if address don't have "on-the-fly" txes (txes with 0 conf) - then balances should be equal. This what I've seen on S3 notary nodes address balances. These addresses are "abandoned", i.e. don't have any active txes right now, and balances are equal for them.

**Review by DeckerSU — APPROVED** [review]
LGTM, as we have decided to merge all major changes into the `main` branch. However, additional and more comprehensive tests will need to be carried out before we can tag the `alpha` release.

**Comment by dimxy** [comment]
closes #6

**Pull request merged**